### PR TITLE
2103 link contact to patient

### PR DESCRIPTION
**Pull request by clemens-msupply**
Part of #2103

# 👩🏻‍💻 What does this PR do? 

Add a "Link contact to mSupply patient" modal to the contact tracing page.

image

image

This is the very base implementations and has a couple of shortcomings/ could be improved:
- For some reason search input fields loose focus after search which is super annoying, any ideas why?
- Not able to unlink patient
- Not able to create new patient (you first have to create the patient and the link it)
- UI design: e.g. button into top right?
- Better visualization needed if patient is linked or not
- Use same UI or reuse components from the patient creation modal
- Fetch patient from central when not found locally (see point above)

# Testing 

- Create a contact trace entry
- Optional: Enter some contact details under "Family Partner information"
- Click the new "Link to patient" button
- If contact details have been entered the search fields should be prepopulated
- Search for a patient and select it
- The icon of the "Link to patient" button should be a check now, indicating that the patient is linked (better visualization needed)
- Relink to different patient
- Save and reload

### CODE REVIEW

File: server/graphql/programs/src/queries/contact_trace.rs [hunk 1]
```diff
@@ -169,17 +169,18 @@ impl ContactTraceNode {
         &self.trace_row().patient_id
     }
 
-    pub async fn patient(&self, ctx: &Context<'_>) -> Result<NameNode> {
-        let loader = ctx.get_loader::<DataLoader<NameByIdLoader>>();
+    pub async fn patient(&self, ctx: &Context<'_>) -> Result<PatientNode> {
+        let loader = ctx.get_loader::<DataLoader<PatientLoader>>();
```
Comment (clemens-msupply) [inline server/graphql/programs/src/queries/contact_trace.rs:173]: Noticed, that this was wrong...

File: client/packages/system/src/ContactTrace/DetailView/DetailView.tsx [hunk 2]
```diff
@@ -140,15 +140,19 @@ export const DetailView: FC<DetailViewProps> = ({
         </span>
       );
     }
-  }, [contactData]);
+  }, [contactData, setSuffix]);
```
Comment (clemens-msupply) [inline client/packages/system/src/ContactTrace/DetailView/DetailView.tsx:143]: From react-hooks/exhaustive-deps. There are more deps missing, but they have issues as well, i.e. they re-render too often leading to infinite loops. The linter revealed it nicely, but I haven't fixed it yet...

**Comment by mark-prins** [comment]
while I think of it - is there any plan to change the contact edit page to a modal or similar?
I was momentarily confused because I'm on a page which does not match the navigation

**Comment by clemens-msupply** [comment]
> while I think of it - is there any plan to change the contact edit page to a modal or similar? I was momentarily confused because I'm on a page which does not match the navigation

What do you mean it works the same as for encounters.

There was the idea to use a modal. But I don't think its a good idea. Mainly because of this change, having a link button in a modal does work very well IMHO. You might also want to add a sidebar for comments, etc. as has been done for encounters...

**Comment by mark-prins** [comment]
> What do you mean it works the same as for encounters.

Encounters has an entry in the left hand navigation; the contact page has no indication of placement, other than the breadcrumb. The difference is troubling me.

The main problem I have though, is that the search form is re-rendering and therefore losing focus. If I type `aa` for example, the second `a` is capture by a global handler and I'm navigated to the admin section. Am just putting in a suggestion to fix that now..

**Review by mark-prins — CHANGES_REQUESTED** [review]
a suggested fix for the search inputs is in [this branch](https://github.com/openmsupply/open-msupply/compare/2103-link-contact-to-patient...2103-suggestion)

the inputs were re-rendering due to the changes in state, so I've lifted the components out. I then had an issue with the captured enclosure in the SearchBar due to the debounce, so had to split the state into setPatch and setFilter. It's working nicely now 

I'm confused with intended functionality - I have a contact, I've linked a patient - and now the name of the contact (in the contact list) is different to the name shown when you edit the contact (I've linked a different patient name).

Two questions:
- should the patient and contact names differ, what should show? both names? the linked patient should override?
- I still have a 'ink to patient' button showing, even though I've linked. Could there be an indication that this contact is already linked?

File: client/packages/system/src/ContactTrace/DetailView/LinkPatientModal.tsx [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+    (filter.firstName?.length ?? 0) > 0 ||
+    (filter.lastName?.length ?? 0) > 0 ||
+    (filter.identifier?.length ?? 0) > 0;
+  const { data: localSearchData } = usePatient.document.list(
+    {
+      filterBy: {
+        firstName: filter.firstName ? { like: filter.firstName } : null,
+        lastName: filter.lastName ? { like: filter.lastName } : null,
+        identifier: filter.identifier ? { like: filter.identifier } : null,
+      },
```
[the remark this answers is not in the document]
Comment (clemens-msupply) [reply client/packages/system/src/ContactTrace/DetailView/LinkPatientModal.tsx:100]: Thanks!

File: client/packages/system/src/ContactTrace/DetailView/LinkPatientModal.tsx [hunk 4]
```diff
@@ -0,0 +1,201 @@
...
+};
+
+const ModalContent: FC<ModalContentProps> = ({
+  documentData,
+  onPatientLinked,
+  hideDialog,
+}) => {
+  const t = useTranslation('dispensary');
+  const { localisedDate } = useFormatDateTime();
+  const onFilterChanged = useCallback(() => {}, []);
```
[the remark this answers is not in the document]
Comment (clemens-msupply) [reply client/packages/system/src/ContactTrace/DetailView/LinkPatientModal.tsx:83]: No this was by mistake. Thanks for removing it!

**Comment by clemens-msupply** [comment]
> the inputs were re-rendering due to the changes in state, so I've lifted the components out. I then had an issue with the captured enclosure in the SearchBar due to the debounce, so had to split the state into setPatch and setFilter. It's working nicely now

Thanks for fixing it that saved me some time!

> I'm confused with intended functionality - I have a contact, I've linked a patient - and now the name of the contact (in the contact list) is different to the name shown when you edit the contact (I've linked a different patient name).
> 
> Two questions:
> 
>     * should the patient and contact names differ, what should show? both names? the linked patient should override?

Good point, I would say the linked patient takes precedence. Or maybe show both if they differ, e.g. "Chris (recorded as Cris)"? Or show the recorded name on hover, if different?

> 
>     * I still have a 'ink to patient' button showing, even though I've linked. Could there be an indication that this contact is already linked?

Yes there is but it is too subtle. The icon of the button changes. I stopped with it because I wasn't sure if I should put the link button into the top AppBarButtons instead and then have a label next to the linked patient.

Maybe we could have a quick chat in person about the design.

**Comment by mark-prins** [comment]
👋  sorry @clemens-msupply  - I forgot to upload this when reviewing earlier!

Screenshot 2023-09-08 at 2 30 07 PM

I thought the smiley face and exclamation mark were not appropriate on this search result, where we are trying to find a match. Perhaps worth changing the translation string for this one.

**Comment by clemens-msupply** [comment]
I made some changes:

Modal now showes the linked patient:
image
image

You can also unlink the patient in the modal. Furthermore, you now have to confirm (ok button) before linking/unlinking the patient.

If the recorded name mismatches the linked patient this is shown using an InfoTooltipIcon:
image

There are still some annoying and visible re-renders but thought it might be something for R&D day to follow this probably (if not found earlier)...

File: client/packages/common/src/ui/icons/Link.tsx [hunk 5]
```diff
@@ -0,0 +1,24 @@
+import React from 'react';
+import SvgIcon, { SvgIconProps } from '@mui/material/SvgIcon';
+
+export const LinkIcon = (props: SvgIconProps): JSX.Element => {
```
Comment (mark-prins) [inline client/packages/common/src/ui/icons/Link.tsx:4]: this looks great! can you also add to the storybook please? (`Icon.stories.tsx`)

**Review by mark-prins — COMMENTED** [review]
🎉  That looks fantastic! thanks for the changes.

I have two more suggestions.. if you'll bear with me.

Firstly, there's an issue with the modal reloading when you click on a patient. That clears the filter which is a bit strange, so I've made a [suggestion branch](https://github.com/openmsupply/open-msupply/compare/2103-link-contact-to-patient...2103-link-contact-to-patient-suggestion) to resolve that.

As part of the change, I've highlighted the selected row like this:

Screenshot 2023-09-13 at 3 10 16 PM

I was finding it hard to see what I'd clicked on as nothing much changed.

Second thing - The info icon stands out, being the orange primary colour. I tried with 'gray.main' instead and I think that looks better:

Screenshot 2023-09-13 at 3 19 04 PM

I've added the change as a separate commit to the suggestion branch

File: client/packages/common/src/ui/icons/Link.tsx (hunk 5 above)
Comment (clemens-msupply) [reply client/packages/common/src/ui/icons/Link.tsx:4]: Thanks done

**Comment by clemens-msupply** [comment]
Thanks for your changes!! There is still a reload problem when clicking a patient, but I will not follow up on this one right now...

I did one fix on top of your branch. There was some stale query data in a hook which caused the row highlighting to not work in some cases...

**Review by mark-prins — APPROVED** [review]
👍 It's a beautiful thing - thanks again @clemens-msupply!

**Pull request merged**